src/utils/chunks/types.rs

```diff
@@ -0,0 +1,70 @@
+//! Contains data types used in the chunk upload process.
+
+use anyhow::Result;
+use sha1_smol::Digest;
+
+use crate::utils::chunks::Chunk;
+use crate::utils::fs;
+
+/// Information returned by `assemble_difs` containing flat lists of incomplete
+/// objects and their missing chunks.
+pub type MissingObjectsInfo<'m, T> = (Vec<&'m ChunkedObject<T>>, Vec<Chunk<'m>>);
+
+/// Chunked arbitrary data with computed SHA1 checksums.
+pub struct ChunkedObject<T>
+where
+    T: AsRef<[u8]>,
+{
+    /// Original object
+    object: T,
+
+    /// SHA1 checksum of the entire object
+    checksum: Digest,
+
+    /// SHA1 checksums of all chunks
+    chunks: Vec<Digest>,
+
+    /// Size of a single chunk
+    chunk_size: usize,
+}
+
+impl<T> ChunkedObject<T>
+where
+    T: AsRef<[u8]>,
+{
+    /// Creates a new `ChunkedObject` from the given object, using
+    /// the given chunk size.
+    pub fn from(object: T, chunk_size: usize) -> Result<Self> {
+        let (checksum, chunks) = fs::get_sha1_checksums(object.as_ref(), chunk_size)?;
+        Ok(Self {
+            object,
+            checksum,
+            chunks,
+            chunk_size,
+        })
+    }
+
+    /// Returns the SHA1 checksum of the entire object.
+    pub fn checksum(&self) -> Digest {
+        self.checksum
+    }
+
+    /// Returns the original object.
+    pub fn object(&self) -> &T {
+        &self.object
+    }
+
+    /// Returns the SHA1 checksums of each chunk, in order.
+    pub fn chunks(&self) -> &[Digest] {
+        &self.chunks
+    }
+
+    /// Returns an iterator over the chunks of the object.
+    pub fn iter(&self) -> impl Iterator<Item = Chunk<'_>> {
```

Comment:
This could use a clearer name and documentation.

Reply:
Maybe this should be called `chunks` and `chunks` should be `chunk_digests` or something along those lines.

Reply:
@loewenheim I settled for calling this one `iter_chunks`, and I renamed `chunks` to `chunk_hashes`. I'll update the PR once I have addressed everything locally